controllers/csidriver.go

```diff
@@ -0,0 +1,98 @@
+/*
+Copyright 2022 Red Hat OpenShift Data Foundation.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	v1k8scsi "k8s.io/api/storage/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	v1 "k8s.io/client-go/kubernetes/typed/storage/v1"
+)
+
+type csiDriver struct {
+	name           string
+	fsGroupPolicy  string
+	attachRequired bool
+	mountInfo      bool
+	csiDriver      *v1k8scsi.CSIDriver
+	csiClient      v1.CSIDriverInterface
+}
+
+// createCSIDriverInfo Registers CSI driver by creating a CSIDriver object
+func (d csiDriver) createCSIDriverInfo(ctx context.Context) error {
+	// Create CSIDriver object
+	csiDriver := &v1k8scsi.CSIDriver{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: d.name,
+		},
+		Spec: v1k8scsi.CSIDriverSpec{
+			AttachRequired: &d.attachRequired,
+			PodInfoOnMount: &d.mountInfo,
+		},
+	}
+	if d.fsGroupPolicy != "" {
+		policy := v1k8scsi.FSGroupPolicy(d.fsGroupPolicy)
+		csiDriver.Spec.FSGroupPolicy = &policy
+	}
+	driver, err := d.csiClient.Get(ctx, d.name, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			_, err = d.csiClient.Create(ctx, csiDriver, metav1.CreateOptions{})
+		}
+		return err
+	}
+
+	// As FSGroupPolicy field is immutable, should be set only during create time.
+	// if the request is to change the FSGroupPolicy, we are deleting the CSIDriver object and creating it.
+	if driver.Spec.FSGroupPolicy != nil && csiDriver.Spec.FSGroupPolicy != nil && *driver.Spec.FSGroupPolicy != *csiDriver.Spec.FSGroupPolicy {
+		return d.reCreateCSIDriverInfo(ctx)
```

Comment:
Any reason to not just call `Delete()` here and use `CreateOrUpdate()` below?

Reply:
reCreateCSIDriverInfo is doing the same, it's doing delete and Create.